manager/allocator/network/allocator.go

```diff
@@ -0,0 +1,869 @@
+	// only place we need it, let's init the drivers. If this fails, it means
+	// the whole system is megascrewed
+	for _, init := range initializers {
+		if err := reg.AddDriver(init.ntype, init.fn, nil); err != nil {
+			panic(fmt.Sprintf("reg.AddDriver returned an error: %v", err))
+		}
+	}
+
+	// then, initialize the IPAM drivers
+	if err := initIPAMDrivers(reg); err != nil {
+		panic(fmt.Sprintf("initIPAMDrivers returned an error: %v", err))
+	}
+	return &allocator{
+		services:          map[string]*api.Service{},
+		nodeLocalNetworks: map[string]*api.Network{},
+		port:              port.NewAllocator(),
+		ipam:              ipam.NewAllocator(reg),
+		driver:            driver.NewAllocator(reg),
+	}
+}
+
+// Restore takes slices of the object types managed by the network allocator
+// and syncs the local state of the Allocator to match the state of the objects
+// provided.
+//
+// If an error occurs during the restore, the local state may be inconsistent,
+// and this allocator should be abandoned>
+func (a *allocator) Restore(networks []*api.Network, services []*api.Service, tasks []*api.Task, nodes []*api.Node) error {
+	// there is a problem with restoring nodes: because node deallocation
+	// depends on network deallocation, it is possible for a network to be
+	// deallocated but the corresponding deallocation of nodes to not occur.
+	// this can lead to a situation on restore where a node has attachments
+	// belonging to no network.
+	//
+	// the simple fix, first off, is to ignore attempts to deallocate an
+	// attachment which has no corresponding network. this allows the caller to
+	// get correct allocator and node state by simply reallocating all nodes
+	// after restoring network state.
+	//
+	// however, Restore will return errors if it attempts to restore an
+	// attachment for which the network is not allocated; specifically,
+	// ipam's Restore will return ErrBadState and no restore any subsequent
+	// (correct) allocations.
+	//
+	// to avoid this case, we'll keep a set of all network IDs that we've
+	// encountered so far. then, when we compile our slice of attachments to
+	// restore, we'll skip any node attachments belonging to a network not in
+	// this list
+	//
+	// because networks cannot be updated, we can be sure that any node
+	// attachment that exists either belongs to a fully allocated network, or
+	// belongs to a deleted network, so we can put even unallocated networks in
+	// this map.
+	//
+	// This issue does not exist with task attachments, because a network
+	// cannot be deleted while it is still in use by tasks.
+	existingNetworks := make(map[string]struct{}, len(networks))
+
+	// find if we have an ingress network in this list. if so, save its ID. we
+	// need it to correctly allocate tasks and services. there should only ever
+	// be 1 ingress network
+	for _, nw := range networks {
+		existingNetworks[nw.ID] = struct{}{}
+
+		if isIngress(nw) {
+			a.ingressID = nw.ID
```

Comment:
Maybe. I want to avoid importing the logging package anywhere deeper into the allocator codebase, and so I'd like to bubble up that information somehow instead of logging down here.

Reply:
+1 for bubbling up.  Out of curiosity, why do you want to avoid importing the logging package?

Reply:
A few reasons:

1. Logging in swarmkit is best accomplished with the `log` package, which propagates logging through contexts. The lower levels of the network allocator have no need for and make no use of context, so logging would have to be accomplished in a more roundabout way, or the API would have to add contexts just for the sole purposes of logging

2. Keeping the import graph clean ensures that the implementation of the network allocator is decoupled from the implementation of the logging system. It also means there is less functionality to test in the network allocator.

3. If we desire later to change the way logging is handled, there is only 1 place to do so (at the top level) instead of many places.

Reply:
Ah that's true about the contexts - forgot about that.